Twister: Sharing built applications across multiple tests

gchwier opened the pull request:
Related to #62649 
and @gopiotr 's PR: #66188 

Added sample to demonstrate sharing build mechanism in samples/subsys/testsuite/shared_app,
try to run with:
```
./scripts/twister -T samples/hello_world -T samples/subsys/testsuite/shared_app -v -p native_sim -p qemu_cortex_m3
```

Changes:
* added `required_images` key to sample/testcase yaml. It is a list of test scenarios that are required to be build before current test. If required image is not ready, then processing of the current test is postponed (received to the pipeline queue). There is also option to use the first required image as an image used during current test (user must set new key `no_own_image` or there is no source (CMakeList.txt) in the current test directory)
* most of the changes are in `runner.py` file. `pipeline` queue is changed from LifoQueue to `deque` (to allow append elements in both sides), `done` is changed from LifoQueue to `dict` (to check if test scenario is ready with constant time).
* in `testplan.py` checked if required images are in scope, if platform is correct etc. Added special filter to mark filtered test scenario with proper comments
* updated unit tests and added new keys to twister doc

This feature does not work with `--subset`.
For QEMU platforms there is no support for `no_own_image` (QEMU_PIPE file is used during build, so to reuse qemu build, one must point to the origin build directory).

All required images must be available from tree (user must give path with `-T` option). If required image is not found in scope, then is skipped.

Usecases:
1) as provided in #62649  with bluetooth samples
2) I want to test DFU and in many cases I can use another image during upgrade (e.g. hello_world or smp_svr) to test downgrade prevention mechanism, wrong keys, wrong encryption used for MCUboot etc. (from the NCS perspective, a lot of tests with B0 bootloader)
3) just to test another sample without copying it. I want to test e.g. smp_svr with pytest scenarios. With one sample, without modifying it and rebuilding, I can run many pytest testcases

gchwier committed:
twister: scripts: implement sharing build mechanism

Added feature to share builds between test scenarios in Twister.
New keys added to test yaml: required_images and no_own_image.

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

gchwier committed:
twister: unittests: align tests with changes for shared apps

Fixed unittests after adding shared apps feature.

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

gchwier committed:
samples: added sample to demonstrate new Twister feature.

Added sample to demonstrate shared app mechanism added
to Twister.

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

gchwier committed:
doc: twister: Added required_images key description.

Added short description of new keys added with shared
app mechanism.

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

hakehuang commented on doc/develop/test/twister.rst:
```diff
@@ -780,6 +780,35 @@ filter: <expression>
+
+        tests:
+          shared.app.example:
+            required_images:
```
both image will build and how to flash them?

gchwier replied:
Flashing is done by Twister as it is now. If the test scenario does not have own image, then build_dir of first required_image is copied to the build_dir of the current test, and is used to flash the board. The other required images are only to use in custom test implementation. For pytest-harness I pass paths to build_dirs of required images. Then one can use those paths to get artefacts, to use them in custom pytest testcases (binaries for upgrades, or to flash another board that is connected to our DUT ... but flashing of second board must be added in custom script)

hakehuang commented:
and the atomic pacakge is other application right?

gchwier commented:
Added checking for circular dependency problem, if in required_images is listed test scenario, which also requires the current app.
In previous update I also disabled support for required images, when `--runtime-artifact-cleanup` is enabled.

LukaszMrugala commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -1015,11 +1021,94 @@ def apply_filters(self, **kwargs):
+            if instance.testsuite.no_own_image or not Path(
+                    Path(instance.testsuite.source_dir) / 'CMakeLists.txt').exists():
+                instance.no_own_image = True
+                # filter out qemu platforms, because for tests must use origin path to QEMU_PIPE,
```
Something is grammatically wrong in this comment, I cannot really comprehend it. Should `for` be here?

gchwier replied:
I removed this comment.

LukaszMrugala commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -1015,11 +1021,94 @@ def apply_filters(self, **kwargs):
+                    Path(instance.testsuite.source_dir) / 'CMakeLists.txt').exists():
+                instance.no_own_image = True
+                # filter out qemu platforms, because for tests must use origin path to QEMU_PIPE,
+                # can be done, but first better to refactor QemuHandler
```
Move `first` to the end of the line

gchwier replied:
removed comment

LukaszMrugala commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -1015,11 +1021,94 @@ def apply_filters(self, **kwargs):
+
+                if req_instance.status:
+                    # check if required image is filtered because is not runnable
+                    if self.options.device_testing and not req_instance.run and \
+                            len(req_instance.filters) == 1 and req_instance.reason == "Not runnable on device":
```
The magic string `reason` comparison seems like a not great solution, but I do not see anything better than that currently.

LukaszMrugala reviewed (changes requested):
Only the comments are a hard no. Two important questions about failed required image reporting. The rest are nitpicks.

aescolar reviewed (commented):
Thanks @gchwier 
some questions and 2 nits
(I have not reviewed the python changes themselves)

aescolar commented:
>  If required image is not ready, then processing of the current test is postponed

Right now, if the requires image does not exist and is not in the plan, the dependent test will just be skipped.
But wouldn't it be more intuitive and user friendly if a test's required images were built/added to the queue if missing from the plan?

As a user, if want to run a test, I would want to just do `twister -s <my_test>` and hope that twister builds the dependencies.
If instead I need to do twister -s <my_test> -s <my_test_required_image_1> -s <my_test_required_image_2> it could get very tedious.

gchwier commented:
@LukaszMrugala @aescolar thank you for review / comments. I am on vacation now, I will reply next week

gchwier commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -1311,51 +1307,108 @@ def add_tasks_to_queue(self, pipeline, build_only=False, test_only=False, retry_
+            instance.status = "skipped"
+            for tc in instance.testcases:
+                tc.status = "skipped"
+            instance.reason = "Required image failed"
```
I wanted to keep it short. This info is displayed on console in runtime, also when verbosity is not enabled. And what if more required builds failed, should we list them all? As you wrote, this info can be found in debug logs. And all required images are built with the same twister execution as the parent image, so you can find (in twister.json or in logs) all failed builds.

gchwier commented on doc/develop/test/twister.rst:
```diff
@@ -780,6 +780,35 @@ filter: <expression>
+            no_own_image: True
+
+    Not supported with options: ``--subset`` or ``--runtime-artifact-cleanup``.
+    Tests on QEMU platforms will be skipped if test does not have an own image.
```
removed this line. In `no_own_image` is mentioned, that it is not supported for Qemu (test will be skipped) so it should be enough

gchwier commented on doc/develop/test/twister.rst:
```diff
@@ -780,6 +780,35 @@ filter: <expression>
+          shared.app.example:
+            required_images:
+              - test.example.first
+              - native_sim:test.example.second
```
Yes, correct.
remember to provide both `-p platform` in twister command to not filter building required image.

gchwier commented:
> > If required image is not ready, then processing of the current test is postponed
> 
> Right now, if the required image does not exist and is not in the plan, the dependent test will just be skipped. But wouldn't it be more intuitive and user friendly if a test's required images were built/added to the queue if missing from the plan?
> 
> As a user, if want to run a test, I would want to just do `twister -s <my_test>` and hope that twister builds the dependencies if missing (and maybe re-trigger an incremental build for those if missing from the plan but present) If instead I need to do twister -s <my_test> -s <my_test_required_image_1> -s <my_test_required_image_2> it could get very tedious.

I agree that it would be more user friendly, but for now, I wanted only provide a "simple" method to be able to reuse other images (and I see that "simple" was only at the beginning). User must provide every sample/path only if want to run a selected sample, when running it with integration mode (`-G`) with `-T tests/ -T samples/`, then all configuration should be available in test plan.
When we want to add automatically all missing configurations, we need to know, where to find them (search ZEPHYR_BASE directory? It can be good idea, however I'd prefer to add it later as an improvement, when this basic solution will be ready).

gchwier commented:
> @gchwier Another question, if I run with `--no-clean -c` I don't get tests which already existed reused(?)

Maybe I do not understand all options in Twister and I do not tested all of them.
`no-clean` should work, I've tried with commands:
`./scripts/twister -T samples/hello_world -T samples/subsys/testsuite/shared_app -p native_sim -vv -c --build-only`
and then
`./scripts/twister -T samples/hello_world -T samples/subsys/testsuite/shared_app -p native_sim -vv --no-clean`

however I found, that `--test-only` does not load all images properly. I must check that.

gchwier commented:
updated with changes to properly add required images when tests are loaded from file (options `--test-only`, `--load-tests`, `--only-failed`)

aescolar commented on doc/develop/test/twister.rst:
```diff
@@ -789,6 +789,34 @@ filter: <expression>
+    to True, then the first image from this list will be used as the test
+    scenario image.
+    All required test scenarios must be available from the tree (given with ``-T`` option).
+    If required test scenario is not available in tree or build is failed,
```
nit
```suggestion
    If a required test scenario is not available in tree or its build failed,
```

aescolar commented on doc/develop/test/twister.rst:
```diff
@@ -789,6 +789,34 @@ filter: <expression>
+    If required test scenario is not available in tree or build is failed,
+    the test will be skipped.
+    If the required scenario is prefixed with a platform name, then it will be
+    build for that platform only.
```
```suggestion
    built for that platform only.
```

aescolar reviewed (dismissed):
2 nits. Overall +1 to the idea

gchwier commented:
Thank you @aescolar for review. I rebased PR and updated twister.rst with your comments.
 > Another question, if I run with --no-clean -c I don't get tests which already existed reused(?)

Now I know what you mean. It could be great to be able to run
`twister --build-only <all_required_images>`
and then:
`twister -T <image_that_requires_another_images> ... --no-clean`
For now it is not possible. You must have every required image in test plan, so you must include it to the command. It is not enough to have it built in twister-out dir.
I will try to add this feature later (some changes are required in `testplan.py` to find image in twister-out dir and to process them in `runner.py`), but in separate PR.

gchwier commented:
Rebased to work with `TwisterStatus`.
@LukaszMrugala Could you please review the changes once again? Thank you!

nashif commented:
@gchwier 

this basically does what I commented on in https://github.com/zephyrproject-rtos/zephyr/issues/62649#issuecomment-1757608037, correct?

gchwier commented:
> this basically does what I commented on in [#62649 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/62649#issuecomment-1757608037), correct?

Yes, correct. This change modifies the scheduler and queue management. Instead of using `LifoQueue`, it now uses `deque` and `dict` (registered with a `BaseManager` to be shared between multiple processes).
* `dict` is used for `done` instances to quickly check if the processing of required testinstances is finished.
* `deque` is used for `pipeline` instances to allow postponing the processing of a task (re-adding the task to the front of the queue) if the required instances have not been processed yet.

aescolar commented:
Oh well, it is really a pitty this has not got anywhere so far, @gchwier @nashif what would need to happen for this to be merged?

gchwier commented:
Rebase is not trivial now (changes related to statuses, ruff etc.). I can do that, however as I understand, tinkering in Twister scheduler and queue management was a blocker here. (https://github.com/zephyrproject-rtos/zephyr/pull/75291#issuecomment-2368545745)
